Source/WebCore/Modules/identity/CredentialRequestCoordinator.cpp

```diff
@@ -28,100 +28,273 @@
+        m_coordinator->setState(PickerState::Idle);
+}
+
+CredentialRequestCoordinator::PickerState CredentialRequestCoordinator::currentState() const
+{
+    return m_state;
+}
+
+bool CredentialRequestCoordinator::canTransitionTo(PickerState newState) const
+{
+    switch (m_state) {
+    case PickerState::Idle:
+        return newState == PickerState::Presenting;
+    case PickerState::Presenting:
+        return newState == PickerState::Aborting || newState == PickerState::Idle;
+    case PickerState::Aborting:
+        return newState == PickerState::Idle;
     }
+    ASSERT_NOT_REACHED();
+    return false;
+}
 
-    if (!document.hasFocus()) {
-        promise.reject(Exception { ExceptionCode::NotAllowedError, "The document is not focused."_s });
+void CredentialRequestCoordinator::setState(PickerState newState)
+{
+    if (m_state == newState || !canTransitionTo(newState))
         return;
+    m_state = newState;
+}
+
+void CredentialRequestCoordinator::setCurrentPromise(CredentialPromise&& promise)
+{
+    ASSERT(!m_currentPromise.has_value());
+    m_currentPromise = WTFMove(promise);
+}
+
+CredentialPromise* CredentialRequestCoordinator::currentPromise()
+{
+    return m_currentPromise ? &m_currentPromise.value() : nullptr;
+}
+
+bool CredentialRequestCoordinator::presentPicker(CredentialPromise&& promise, DigitalCredentialsRequestData&& requestData, RefPtr<AbortSignal> signal)
+{
+    if (m_state != PickerState::Idle) {
+        LOG(DigitalCredentials, "A credential picker operation is already in progress");
+        promise.reject(Exception {
+            ExceptionCode::InvalidStateError,
+            "A credential picker operation is already in progress."_s });
+        return false;
+    }
+
+    if (!m_page) {
+        promise.reject(ExceptionCode::InvalidStateError, "Page no longer valid."_s);
+        return false;
     }
 
-    if (document.visibilityState() != VisibilityState::Visible) {
-        promise.reject(Exception { ExceptionCode::NotAllowedError, "The document is not visible."_s });
+    if (signal) {
+        if (signal->aborted()) {
+            LOG(DigitalCredentials, "AbortSignal was already aborted before presenting the credential picker");
+            promise.rejectType<IDLAny>(signal->reason().getValue());
+            return false;
+        }
+
+        auto weakThis = WeakPtr { *this };
+        signal->addAlgorithm([weakThis, signal = RefPtr { signal }](JSC::JSValue reason) {
+            if (!weakThis)
+                return;
+            LOG(DigitalCredentials, "Credential picker was aborted by AbortSignal");
+            weakThis->abortPicker(reason);
+        });
+    }
+
+    setState(PickerState::Presenting);
+    setCurrentPromise(WTFMove(promise));
+
+    auto weakThis = WeakPtr { *this };
+    m_client->showDigitalCredentialsPicker(
+        requestData,
+        [weakThis, signal = RefPtr { signal }](Expected<DigitalCredentialsResponseData, ExceptionData>&& responseOrException) {
```

Comment:
```suggested
        [weakThis = WeakPtr { *this }, signal](Expected<DigitalCredentialsResponseData, ExceptionData>&& responseOrException) {
```

no need for intermediary weakThis variable. signal can be copy captured already, it's a RefPtr, no need for `RefPtr { ... }` again